framework/src/preference/preference_set.c

```diff
@@ -0,0 +1,300 @@
+/****************************************************************************
+ *
+ * Copyright 2019 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <unistd.h>
+#include <debug.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+#include <debug.h>
+#include <sys/types.h>
+#include <preference/preference.h>
+
+#include "preference_internal.h"
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static int preference_fs_write_key(char *path, preference_data_t *data)
+{
+	int fd;
+	int ret;
+
+	if (path == NULL || data == NULL) {
+		prefdbg("Invalid parameter\n");
+		PREFERENCE_FREE(path);
+		return PREFERENCE_INVALID_PARAMETER;
+	}
+
+	fd = open(path, O_WRONLY | O_CREAT, 0666);
+	if (fd < 0) {
+		prefdbg("open fail %d\n", errno);
+		PREFERENCE_FREE(path);
+		return PREFERENCE_IO_ERROR;
+	}
+
+	/* Write the type of data value */
+	ret = write(fd, (void *)&(data->type), sizeof(int));
+	if (ret < 0) {
+		ret = PREFERENCE_IO_ERROR;
+		goto errout;
+	}
+
+	/* Write the length of data value */
+	ret = write(fd, (void *)&(data->len), sizeof(int));
+	if (ret < 0) {
+		ret = PREFERENCE_IO_ERROR;
+		goto errout;
+	}
+
+	/* Write the data value */
+	switch (data->type) {
+	case PREFERENCE_TYPE_INT:
+	case PREFERENCE_TYPE_BOOL:
+		prefvdbg("[WRITE] int type %d\n", data->value.i);
+		ret = write(fd, (void *)&(data->value.i), data->len);
+		break;
+	case PREFERENCE_TYPE_DOUBLE:
+		prefvdbg("[WRITE] double type %f\n", data->value.d);
+		ret = write(fd, (void *)&(data->value.d), data->len);
+		break;
+	case PREFERENCE_TYPE_STRING:
+		prefvdbg("[WRITE] string type %s\n", data->value.s);
+		ret = write(fd, (void *)data->value.s, data->len);
+		PREFERENCE_FREE(data->value.s);
+		break;
+	default:
+		prefdbg("invalid type %d\n", data->type);
+		ret = PREFERENCE_INVALID_PARAMETER;
+		goto errout;
+	}
+
+	if (ret < 0) {
+		ret = PREFERENCE_IO_ERROR;
+		goto errout;
+	}
+	close(fd);
+	prefvdbg("Write Key Success : %s\n", path);
```

Comment:
We don't need to unlink for successful finishing?

Reply:
yes, preference value is stored in file.